velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp

```diff
@@ -542,8 +542,485 @@ struct ComplexTypeExtractor {
+      std::optional<V> value,
+      Compare& comparator) {
+    if (base.heapValues.size() < base.n) {
+      auto comparePosition = writeComplex(decodedCompare, compareIndex);
+      addToAccumulator(comparePosition, value, comparator);
+    } else {
+      const auto& topPair = base.heapValues.front();
+      if (comparator.compare(topPair, decodedCompare, compareIndex)) {
+        auto comparePosition = writeComplex(decodedCompare, compareIndex);
+        std::pop_heap(
+            base.heapValues.begin(), base.heapValues.end(), comparator);
+        base.heapValues.pop_back();
+        addToAccumulator(comparePosition, value, comparator);
+      }
+    }
+  }
+
+  /// Extract all values from 'heapValues' into 'rawValues' and 'rawValueNulls'
+  /// buffers. The heap remains unchanged after the call.
+  void extractValues(
+      FlatVector<V>& values,
+      vector_size_t offset,
+      Compare& comparator) {
+    std::sort_heap(base.heapValues.begin(), base.heapValues.end(), comparator);
+    // Add heap elements to values in ascending order.
+    for (int64_t i = 0; i < base.heapValues.size(); ++i) {
+      const auto& pair = base.heapValues[i];
+      extractValue(pair, values, offset + i);
+    }
+    std::make_heap(base.heapValues.begin(), base.heapValues.end(), comparator);
+  }
+
+  /// Moves all pairs of (comparison, value) from 'heapValues' into
+  /// 'rawComparisons', 'rawValues' and 'rawValueNulls' buffers. The heap
+  /// remains unchanged after the call.
+  void extractPairs(
+      BaseVector& compares,
+      FlatVector<V>& values,
+      vector_size_t offset,
+      Compare& comparator) {
+    std::sort_heap(base.heapValues.begin(), base.heapValues.end(), comparator);
+    // Add heap elements to compares and values in ascending order.
+    for (int64_t i = 0; i < base.heapValues.size(); ++i) {
+      const auto& pair = base.heapValues[i];
+      extractCompare(pair, compares, offset + i);
+      extractValue(pair, values, offset + i);
+    }
+    std::make_heap(base.heapValues.begin(), base.heapValues.end(), comparator);
+  }
+
+ private:
+  std::optional<StringView> writeString(
+      const std::optional<StringView>& value) {
+    if (!value.has_value()) {
+      return std::nullopt;
+    }
+    return valueSet.write(*value);
+  }
+
+  HashStringAllocator::Position writeComplex(
+      DecodedVector& decoded,
+      vector_size_t index) {
+    HashStringAllocator::Position position;
+    valueSet.write(*decoded.base(), decoded.index(index), position);
+    return position;
+  }
+
+  void addToAccumulator(
+      C comparePosition,
+      const std::optional<V>& value,
+      Compare& comparator) {
+    base.heapValues.push_back(
+        std::make_pair(comparePosition, writeString(value)));
+    std::push_heap(base.heapValues.begin(), base.heapValues.end(), comparator);
+  }
+
+  void freePair(typename BaseType::Heap::const_reference pair) {
+    valueSet.free(pair.first.header);
+    valueSet.free(*pair.second);
```

Comment:
How this work when pair.second is an std::nullopt?

Reply:
Fixed and ut covered.